File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    
    let p1 = headA;
    let p2 = headB;

    while(p1 != null && p2 != null) {
        if(p1 == p2) return p1;
        if(p1.next == null && p2.next == null) break;
        p1 = p1.next == null ? headB : p1.next;
        p2 = p2.next == null ? headA : p2.next;
    }

    return null;

};

File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode a1 = headA;
        ListNode a2 = headB;
        while(a1 != null && a2 != null ) {
            if(a1 == a2) {
                return a1;
            }
            a1 = a1.next;
            a2 = a2.next;
            if(a1 == a2 ) {
                return a1;
            }
            if(a1 == null) {
                a1 = headB;
            }
            if(a2 == null) {
                a2 = headA;
            }
        }
        return null;
    }
}